executor/aggfuncs/func_max_min.go

```diff
@@ -14,121 +14,133 @@
 package aggfuncs
 
 import (
-	"container/heap"
 	"unsafe"
 
+	"github.com/pingcap/errors"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/stringutil"
 )
 
-type maxMinHeap struct {
-	data    []interface{}
-	varSet  map[interface{}]int64
-	isMax   bool
-	cmpFunc func(i, j interface{}) int
+// NewDeque inits a new Deque
+func NewDeque(isMax bool, cmpFunc func(i, j interface{}) int) *Deque {
+	return &Deque{[]Pair{}, isMax, cmpFunc}
 }
 
-func newMaxMinHeap(isMax bool, cmpFunc func(i, j interface{}) int) *maxMinHeap {
-	h := &maxMinHeap{
-		data:    make([]interface{}, 0),
-		varSet:  make(map[interface{}]int64),
-		isMax:   isMax,
-		cmpFunc: cmpFunc,
-	}
-	return h
+// Deque is an array based double end queue
+type Deque struct {
+	Items   []Pair
+	IsMax   bool
+	cmpFunc func(i, j interface{}) int
 }
 
-func (h *maxMinHeap) Len() int { return len(h.data) }
-func (h *maxMinHeap) Less(i, j int) bool {
-	if h.isMax {
-		return h.cmpFunc(h.data[i], h.data[j]) > 0
-	}
-	return h.cmpFunc(h.data[i], h.data[j]) < 0
+// PushFront pushes Idx and Item(wrapped in Pair) to the front of Deque
+func (d *Deque) PushFront(idx uint64, item interface{}) {
+	d.Items = append([]Pair{{item, idx}}, d.Items...)
 }
-func (h *maxMinHeap) Swap(i, j int) { h.data[i], h.data[j] = h.data[j], h.data[i] }
 
-func (h *maxMinHeap) Push(x interface{}) {
-	h.data = append(h.data, x)
-}
-func (h *maxMinHeap) Pop() interface{} {
-	old := h.data
-	n := len(old)
-	x := old[n-1]
-	h.data = old[0 : n-1]
-	return x
+// PushBack pushes Idx and Item(wrapped in Pair) to the end of Deque
+func (d *Deque) PushBack(idx uint64, item interface{}) {
+	d.Items = append(d.Items, Pair{item, idx})
 }
 
-func (h *maxMinHeap) Reset() {
-	h.data = h.data[:0]
-	h.varSet = make(map[interface{}]int64)
-}
-func (h *maxMinHeap) Append(val interface{}) {
-	h.varSet[val]++
-	if h.varSet[val] == 1 {
-		heap.Push(h, val)
+// PopFront pops an Item from the front of Deque
+func (d *Deque) PopFront() (interface{}, error) {
```

Comment:
The function doesn't need return res..
(And the type should be `Pair`, not inferface{})